### ISSUE TITLE
FEATURE: Admins can see when a subject area was last updated

### ISSUE THREAD

**Issue by pconrad** [ucsb-cs156/proj-courses#27]
# User Story

* As an admin
* I can see when a given subject area/quarter combination was last updated
* So that I can know which subject areas have/have not been updated

# Discussion

At the moment, many of the searches use the mongo db collections as the basis of the search instead of the live API. There are several reasons for this:

* The main reason is that the live API does not support doing searches across multiple quarters
* A secondary reason is that the data for past quarters doesn't change, so it's more efficient to cache it than repeatedly ask the API for it

However, the admin must actually run the job to update the data from past quarters and it can be challenging to know which data has and has not been updated.

The purpose of this epic is to provide a way for the admin to be able to easily see, for any given subjectArea and/or quarter, whether the data has been retrieved, and if so, how recently. 

The Epic described below outlines one way of achieving this, but there may be others (including using the last updated timestamps that already exist in the mongodb collections.  You are encouraged to at the proposed design, and if it seems reasonable, implement it; but if you see an easier way to accomplish the same goal, please discuss it with the staff.  Who knows, your idea might be better!

One more note: eventually, we would like to be updating the mongodb database each time a live query is done on the main search page (which uses the API directly), and possibly even move to a system where we always use the mongo db unless the data is missing or stale, in which case we use the API and then update the mongodb collection.

Stale might be defined as "older than n minutes" where n might be 5 or 10.  We might then add a stipulation that data for past quarters is never considered stale if it is present.

# Note: some parts of this may already have been implemented

One of your first tasks will be to see how much of this epic has already been implemented by students in previous quarters and pick up where they left off.

# Current proposal

The proposal is to provide a table that shows, for each subject area and quarter that has been updated by one of the "Update Course Info" jobs, when it was most recently updated. 

The data behind this table should be a MongoDB collection rather than a table in the SQL database; this is a kind of "metadata" for the MondoDB courses collection (see example), so it should be stored in an adjacent collection.  We propose the name `updates`, and a JSON object with this structure:

Indeed, this table exist:
* Document: `Update.java`: <https://github.com/ucsb-cs156/proj-courses/blob/main/src/main/java/edu/ucsb/cs156/courses/documents/Update.java>
* Collection: `UpdateCollection.java` <https://github.com/ucsb-cs156/proj-courses/blob/main/src/main/java/edu/ucsb/cs156/courses/collections/UpdateCollection.java>

```json
{
  "subject_area" : "CMPSC",
  "quarter_yyyyq": "20234",
  "last_update": "2023-05-13T10:22:35"
}
```

Screenshot of existing courses collection on https://cloud.mongodb.com: 
<img width="996" alt="image" src="https://github.com/ucsb-cs156/proj-courses/assets/1119017/8b2c8e1c-3c6d-4911-9d85-67d706536310">

Dividing this up into smaller issues might look like this (each in it's own PR).  You may collect some of these together in one PR, but beware the danger of PRs that are too big; it is advisable to not combine more than two of these in a single PR.

* Issue: Make a backend endpoints for a GET query `/api/updates?subjectArea=CMPSC&yyyyq=20244` that takes two parameters, subject area and quarter, and returns entries in the table that match the subject area and quarter given, sorted in order from most recent to least recent, and paged with 10 entries per page.
* In the job `src/main/java/edu/ucsb/cs156/courses/jobs/UpdateCourseDataJob.java` add code that upserts a document for the combination of subject-area code and quarter that updates the time last updated ([see explanation of upsert](https://www.geeksforgeeks.org/upsert-in-mongodb/).  This should be done at the end of the update process, only if it was successful, i.e. after each subject area is updated for that particular quarter.
* In parallel with all of the above, make a fixtures document in the frontend code showing sample JSON for the updates (see sample JSON above).  This can be a separate PR, or combined with PRs for the features below.
* Make a frontend component for an updates table based on the fixtures, and make a storybook entry for the table.
* Make a page and link to it from the admin menu with the option "Updates".  This pages should have a SingleSubjectDropdown, and SingleQuarterDropdown, and a "show" button.  When the user clicks "show", it will perform a backend query and load the table with the results.

### Acceptance Criteria

- [ ] There is a way for an admin to see when any given subject area, and/or quarter, was last updated. 

### Implementation Todos

Frontend:
- [ ] There is a table component for the updates collection
- [ ] There is a new page for displaying most recent updates that incorporates all three of the components above.

Backend:
- [ ] There is a collection for updates
- [ ] The jobs that update course data update this new collection
- [ ] There is a GET endpoint for this collection that supports queries by subject area and quarter.
